src/Compiler/Driver/ReuseTcResults/ReuseTcResultsDriver.fs

```diff
@@ -0,0 +1,124 @@
+    let writeThisCompilationReferences = writeThisTcData referencesFilePath
+
+    let getContentHash fileName =
+        use stream = FileSystem.OpenFileForReadShim fileName
+        stream.ReadAllBytes() |> Md5Hasher.computeHash |> BitConverter.ToString
+
+    let getThisCompilationCmdLine args = args
+
+    // maybe split into two things?
+    let getThisCompilationGraph inputs =
+        let sourceFiles =
+            inputs
+            |> Seq.toArray
+            |> Array.mapi (fun idx (input: ParsedInput) ->
+                {
+                    Idx = idx
+                    FileName = input.FileName
+                    ParsedInput = input
+                })
+
+        let filePairs = FilePairMap sourceFiles
+        let graph, _ = DependencyResolution.mkGraph filePairs sourceFiles
+
+        let list = List<string>()
+
+        for KeyValue(idx, _) in graph do
+            let hash = getContentHash sourceFiles[idx].FileName
+            list.Add($"%i{idx}[\"{hash}\"]")
+
+        for KeyValue(idx, deps) in graph do
+            for depIdx in deps do
+                list.Add($"%i{idx} --> %i{depIdx}")
+
+        list.ToArray()
+
+    let getThisCompilationReferences =
+        List.map (fun (r: AssemblyReference) -> r.Text, getContentHash r.Text)
+        >> List.map (fun (name, hash) -> $"{name}: {hash}")
+        >> List.toArray
+
+    member _.TryReuseTcResults inputs =
+        if FileSystem.DirectoryExistsShim tcDataFolderName then
+            use _ = Activity.start Activity.Events.reuseTcResultsCachePresent []
+
+            let prevCompilationCmdLine = getPrevCompilationCmdLine ()
+            let thisCompilationCmdLine = getThisCompilationCmdLine tcConfig.cmdLineArgs
+
+            if prevCompilationCmdLine = thisCompilationCmdLine then
+
+                let prevCompilationReferences = getPrevCompilationReferences ()
+                let thisCompilationReferences = getThisCompilationReferences tcConfig.referencedDLLs
+
+                if prevCompilationReferences = thisCompilationReferences then
+
+                    let prevCompilationGraph = getPrevCompilationGraph ()
+                    let thisCompilationGraph = getThisCompilationGraph inputs
+
+                    if prevCompilationGraph = thisCompilationGraph then
+                        use _ = Activity.start Activity.Events.reuseTcResultsCacheHit []
+
+                        () // do nothing, yet
+                    else
+                        use _ = Activity.start Activity.Events.reuseTcResultsCacheMissed []
+
+                        writeThisCompilationGraph thisCompilationGraph
+                else
+                    use _ = Activity.start Activity.Events.reuseTcResultsCacheMissed []
+
+                    writeThisCompilationReferences thisCompilationReferences
+            else
+                use _ = Activity.start Activity.Events.reuseTcResultsCacheMissed []
+
+                writeThisCompilationCmdLine thisCompilationCmdLine
+        else
```

Comment:
Can cognitive load be reduced here somehow? Not to have a bunch of nested conditions?

Reply:
Okay that looks different now.